File: CORE/src/main/java/com/variant/core/test/jdbc/EventReader.java
package com.variant.core.test.jdbc;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;

import org.apache.commons.collections4.Predicate;

import com.variant.core.impl.VariantCore;
import com.variant.core.jdbc.JdbcAdapter;
import com.variant.core.jdbc.JdbcService;

/**
 * Read events from a JDBC event persister.
 * Tests only
 * 
 * @author Igor
 *
 */
public class EventReader {

	private JdbcService jdbcService = null;

	/**
	 * 
	 * @param variant
	 */
	public EventReader(VariantCore variant) {
		jdbcService = new JdbcService(variant);
	}
	
	/**
	 * Read events as a collection
	 * @return
	 * @throws ClassNotFoundException
	 * @throws SQLException
	 */
	public Collection<VariantEventFromDatabase> readEvents(final Predicate<VariantEventFromDatabase> filter) 
		throws Exception {
		
		final String SELECT_EVENTS_SQL = 
				"SELECT e.id, e.session_id, e.created_on, e.event_name, e.event_value, p.key, p.value " +
				"FROM events e LEFT OUTER JOIN event_params p ON e.id = p.event_id";

		final String SELECT_EVENT_EXPERIENCES_SQL =
				"SELECT id, event_id, test_name, experience_name, is_control " +
				"FROM event_experiences";
		
		return JdbcAdapter.executeQuery(
			jdbcService.getConnection(), 
			new JdbcAdapter.QueryOperation<Collection<VariantEventFromDatabase>>() {

				@Override
				public Collection<VariantEventFromDatabase> execute(Connection conn) throws SQLException {
					
					// Read events
					Statement stmt = conn.createStatement();
					ResultSet rs = stmt.executeQuery(SELECT_EVENTS_SQL);
					
					// Keep items in a map keyed by event ID for easy access.
					HashMap<Long, VariantEventFromDatabase> eventMap = new HashMap<Long, VariantEventFromDatabase>();
					
					while (rs.next()) {
						Long id = rs.getLong(1);
						VariantEventFromDatabase event = eventMap.get(id);
						if (event == null) {
							event = new VariantEventFromDatabase();
							event.id = id;
							event.sessionId = rs.getString(2).trim(); // Fixed width in DB.
							event.createdOn = rs.getDate(3);
							event.name = rs.getString(4);
							event.value = rs.getString(5);
							event.eventExperiences = new HashSet<EventExperienceFromDatabase>();
							event.params = new HashMap<String,String>();
							eventMap.put(id, event);
						}
						String key = rs.getString(6);
						if (key != null) event.params.put(key, rs.getString(7));
					}
					
					// Read event_experiences
					// Keep items in a map (keyed by event ID) of maps (keyed by by event_experience ID) for easy access.
					HashMap<Long, Map<Long, EventExperienceFromDatabase>> outerMap = new HashMap<Long, Map<Long, EventExperienceFromDatabase>>();
					rs = stmt.executeQuery(SELECT_EVENT_EXPERIENCES_SQL);
					
					while (rs.next()) {
						
						Long eventId = rs.getLong(2);
						Map<Long, EventExperienceFromDatabase> innerMap = outerMap.get(eventId);
						if (innerMap == null) {
							innerMap = new HashMap<Long, EventExperienceFromDatabase>();
							outerMap.put(eventId, innerMap);
						}
						
						long evId = rs.getLong(1);
						EventExperienceFromDatabase ev = innerMap.get(evId);
						if (ev == null) {
							ev = new EventExperienceFromDatabase();
							ev.id = evId;
							ev.eventId = eventId;
							ev.testName = rs.getString(3);
							ev.experienceName = rs.getString(4);
							ev.isControl = rs.getBoolean(5);
							innerMap.put(evId, ev);
						}
						
					}
					
					// Attach event_experiences to events
					for (Map.Entry<Long, Map<Long, EventExperienceFromDatabase>> outerEntry: outerMap.entrySet()) {
						VariantEventFromDatabase event = eventMap.get(outerEntry.getKey());
						for (Map.Entry<Long, EventExperienceFromDatabase> innerEntry: outerEntry.getValue().entrySet()) {
							event.eventExperiences.add(innerEntry.getValue());
						}
					}
					
					// Apply filter
					Iterator<VariantEventFromDatabase> result = eventMap.values().iterator();
					while(result.hasNext()) {
						VariantEventFromDatabase e = result.next();
						if (!filter.evaluate(e)) result.remove();
					}
					
					return eventMap.values();
				}
			}
		);
	}			
	
	/**
	 * Read events as a collection
	 * @return
	 * @throws ClassNotFoundException
	 * @throws SQLException
	 */
	public Collection<VariantEventFromDatabase> readEvents() throws Exception {

		Predicate<VariantEventFromDatabase> noFilter = new Predicate<VariantEventFromDatabase>() {

			@Override
			public boolean evaluate(VariantEventFromDatabase t) {
				return true;
			}
		};
		return readEvents(noFilter);
	}
}

File: SERVER/test/com/variant/server/test/util/EventReader.scala
package com.variant.server.test.util

import com.variant.server.jdbc.JdbcService
import com.variant.server.jdbc.JdbcAdapter
import java.util.Collection
import java.sql.Connection
import org.junit.runners.model.Statement
import java.sql.ResultSet
import java.util.HashMap
import java.util.HashSet
import java.util.Map
import scala.collection.JavaConversions._
import com.variant.server.event.EventWriter
import com.variant.server.jdbc.EventFlusherJdbc


/**
 * Read events written by a JDBC event flusher.
 * Tests only
 * 
 * @author Igor
 *
 */
object EventReader {
   def apply(eventWriter: EventWriter) = new EventReader(eventWriter)
}

/**
 * 
 */
class EventReader (eventWriter: EventWriter) {
   
   val jdbcConnection = eventWriter.flusher.asInstanceOf[EventFlusherJdbc].getJdbcConnection

   /**
	 * Read events as a collection
	 * @return
	 * @throws ClassNotFoundException
	 * @throws SQLException
	 */
	def read(p: TraceEventFromDatabase => Boolean = (_ => true)) = {

		val SELECT_EVENTS_SQL = 
				"SELECT e.id, e.session_id, e.created_on, e.event_name, e.event_value, p.key, p.value " +
				"FROM events e LEFT OUTER JOIN event_params p ON e.id = p.event_id";

		val SELECT_EVENT_EXPERIENCES_SQL =
				"SELECT id, event_id, test_name, experience_name, is_control " +
				"FROM event_experiences";
		
		JdbcAdapter.executeQuery(
			jdbcConnection, 
			new JdbcAdapter.QueryOperation[Collection[TraceEventFromDatabase]]() {

				override def execute(conn: Connection) = {
					
					// Read events
					val stmt = conn.createStatement();
					val eventsRresulSet = stmt.executeQuery(SELECT_EVENTS_SQL);

					// Keep items in a map keyed by event ID for easy access.
					val eventMap = new HashMap[Long, TraceEventFromDatabase]();
					
					while (eventsRresulSet.next()) {
						val id = eventsRresulSet.getLong(1);
						var event = eventMap.get(id);
						if (event == null) {
							event = new TraceEventFromDatabase();
							event.id = id;
							event.sessionId = eventsRresulSet.getString(2).trim(); // Fixed width in DB.
							event.createdOn = eventsRresulSet.getTimestamp(3);
							event.name = eventsRresulSet.getString(4);
							event.value = eventsRresulSet.getString(5);
							event.eventExperiences = new HashSet[EventExperienceFromDatabase]();
							event.params = new HashMap[String,String]();
							eventMap.put(id, event);
						}
						val key = eventsRresulSet.getString(6);
						if (key != null) event.params.put(key, eventsRresulSet.getString(7));
					}
					
					// Read event_experiences
					// Keep items in a map (keyed by event ID) of maps (keyed by by event_experience ID) for easy access.
					val outerMap = new HashMap[Long, Map[Long, EventExperienceFromDatabase]]();
					val eventExperiencesResultSet = stmt.executeQuery(SELECT_EVENT_EXPERIENCES_SQL);
					
					while (eventExperiencesResultSet.next()) {
						
						val eventId = eventExperiencesResultSet.getLong(2);
						var innerMap = outerMap.get(eventId);
						if (innerMap == null) {
							innerMap = new HashMap[Long, EventExperienceFromDatabase]();
							outerMap.put(eventId, innerMap);
						}
						
						val eeId = eventExperiencesResultSet.getLong(1);
						var ee = innerMap.get(eeId);
						if (ee == null) {
							ee = new EventExperienceFromDatabase();
							ee.id = eeId;
							ee.eventId = eventId;
							ee.testName = eventExperiencesResultSet.getString(3);
							ee.experienceName = eventExperiencesResultSet.getString(4);
							ee.isControl = eventExperiencesResultSet.getBoolean(5);
							innerMap.put(eeId, ee);
						}
					}
					
					// Attach event_experiences to events
					for ((eventId, inner) <- outerMap) {
						val event = eventMap.get(eventId)
						for ((eeId, eventExperience) <- inner) {
							event.eventExperiences.add(eventExperience)
						}
					}
					eventMap.values()
				}
			}
		).filter(p)
	}
}
